keras_cv/losses/dice.py

```diff
@@ -0,0 +1,167 @@
+        self,
+        beta=1,
+        from_logits=False,
+        class_ids=None,
+        label_smoothing=0.0,
+        per_sample=False,
+        epsilon=keras.backend.epsilon(),
+        name="dice",
+        **kwargs,
+    ):
+        """Initializes `Dice` instance.
+
+        Args:
+            beta: A float or integer coefficient for balancing the precision
+                and recall. It determines the weight of recall and precision
+                in the combined score. If `beta < 1`, precisoin will doninate;
+                if `beta > 1`, recall will dominate. Default to `1`.
+            from_logits: Whether `y_pred` is expected to be a logits tensor. By
+                default, we assume that `y_pred` encodes a probability distribution.
+                Default to `False`.
+            class_ids: An interger or a lost of intergers within `range(num_classes)`
+                to evaluate the loss. If it's `None`, all classes will beu used to
+                calculate the loss. Default to `None`.
+            label_smoothing: Float in [0, 1]. When > 0, label values are smoothed,
+                meaning the confidence on label values are relaxed. For example, if
+                `0.1`, use `0.1 / num_classes` for non-target labels and
+                `0.9 + 0.1 / num_classes` for target labels. Default to `0.0`.
+            per_sample: If `True`, the loss will be calculated for each sample in
+                batch and then averaged. Otherwise the loss will be calculated for
+                the whole batch. Default to `False`.
+            epsilon: Small float added to dice score to avoid dividing by zero.
+                Default to `keras.backend.epsilon()` or `1e-07`.
+            name: Optional name for the instance.
+                Defaults to 'dice'.
+        """
+        super().__init__(name=name, **kwargs)
+        self.beta = beta
+        self.from_logits = from_logits
+        self.label_smoothing = label_smoothing
+        self.per_sample = per_sample
+        self.epsilon = epsilon
+
+        if class_ids is not None:
+            if isinstance(class_ids, float) or any(
+                isinstance(x, float) for x in class_ids
+            ):
+                raise ValueError(
+                    f"The indices should be int or a list of integer. Got {class_ids}"
+                )
+            elif isinstance(class_ids, int):
+                class_ids = [class_ids]
+
+        self.class_ids = class_ids
+
+    def _smooth_labels(self, y_true, y_pred, label_smoothing):
+        num_classes = tf.cast(tf.shape(y_true)[-1], y_pred.dtype)
+        return y_true * (1.0 - label_smoothing) + (label_smoothing / num_classes)
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.convert_to_tensor(y_pred)
+        y_true = tf.cast(y_true, y_pred.dtype)
+        label_smoothing = tf.convert_to_tensor(self.label_smoothing, dtype=y_pred.dtype)
+
+        if self.from_logits:
+            y_pred = tf.nn.softmax(y_pred)
```

Comment:
@qlzh727 In [MONAI](https://github.com/Project-MONAI/MONAI), they set some arguments regarding the last layer activation function ([HERE](https://github.com/Project-MONAI/MONAI/blob/dev/monai/losses/dice.py#L48-L49)) - which helps to properly transform [logits to probabilities.](https://github.com/Project-MONAI/MONAI/blob/dev/monai/losses/dice.py#L129-L140). But keras doesn't follow this design. Instead, it provides separate loss functions (binary cross-entropy, categorical cross-entropy). 

In the above (lines 116-117), we used only softmax for `logit=True`; not for sigmoid activation. So, to maintain the keras design, what could be the fixes here?

Reply:
I think keras will probably keep the existing keras format. We don't apply any activation in the loss function. User should do this in their model instead. If user would like to apply sigmoid activation, they should change the model to include sigmoid activation, and specify a loss with from_logits=False.